In [1]:
import pandas as pd
path = "Emplid_And_Crse2022.csv"
df = pd.read_csv("Emplid_And_Crse2022.csv", dtype={"Emplid": str})

##### **Data Cleaning & Exploration**

In [2]:
# Quick checks
print(df.shape)
display(df.head())
display(df.columns)

# Clean column names
df.columns = (
    df.columns
    .str.strip()
    .str.replace(" ", "_")
    .str.replace("-", "")
)

# DClean Course_Code column
df["Course_Code"] = df["Course_Code"].astype(str).str.strip()

display(df.columns)

(3457, 14)


,CompletionTermDesc,Emplid,Term Descrshort,Crse ID,Subject,Catalog Nbr,Class Section,Session Descrshort,Crse Component Descrshort,Crse Descr,Unt Earned - Credits,Crse Grade Off - Grade Earned,Course Code,Outside UMSI
0,Winter 2022,03189808,WN 2021,17530,TO,616,1,Scnd 7 Wk,Recitation,Project Mngmt,1.5,B,TO 616,Yes
1,Winter 2022,03189808,WN 2021,40667,SI,681,89,Regular,Seminar,Intern/Field Seminar,6.0,S,SI 681,No
2,Winter 2022,03189808,WN 2021,42091,SI,618,1,Regular,Lecture,Data Man & Analysis,3.0,B+,SI 618,No
3,Winter 2022,03189808,WN 2021,47234,ENTR,510,1,Regular,Lecture,Compensation,3.0,A,ENTR 510,Yes
4,Winter 2022,03189808,WN 2021,47640,SI,507,1,Regular,Lecture,Inter Programming,3.0,A+,SI 507,No


Index(['CompletionTermDesc', 'Emplid', 'Term Descrshort', 'Crse ID', 'Subject',
       'Catalog Nbr', 'Class Section', 'Session Descrshort',
       'Crse Component Descrshort', 'Crse Descr', 'Unt Earned - Credits ',
       'Crse Grade Off - Grade Earned', 'Course Code', 'Outside UMSI'],
      dtype='str')

Index(['CompletionTermDesc', 'Emplid', 'Term_Descrshort', 'Crse_ID', 'Subject',
       'Catalog_Nbr', 'Class_Section', 'Session_Descrshort',
       'Crse_Component_Descrshort', 'Crse_Descr', 'Unt_Earned__Credits',
       'Crse_Grade_Off__Grade_Earned', 'Course_Code', 'Outside_UMSI'],
      dtype='str')

##### **Data Analysis**

In [3]:
# How many unique students are in the dataset?
all_students = set(df["Emplid"].unique())
print(f"Number of Students: {len(all_students)}")

# How many unique courses are in the dataset?
print(f"Number of Unique Courses: {df['Course_Code'].nunique()}")

# Most popular courses
print("Most Popular Courses by Course Code:")
print(df["Course_Code"].value_counts().head(10))

Number of Students: 216
Number of Unique Courses: 292
Most Popular Courses by Course Code:
Course_Code
SI 699    202
SI 501    186
SI 681    167
SI 506    141
SI 539    127
SI 582    114
SI 588    109
SI 622    106
SI 520     96
SI 690     87
Name: count, dtype: int64


##### **MTOP & Capstone Course Count**

In [4]:
df[df["Course_Code"].isin(["SI 697", "SI 698"])]["Emplid"].nunique()
print(f"Number of students who took SI 697 or SI 698: {df[df['Course_Code'].isin(['SI 697', 'SI 698'])]['Emplid'].nunique()}")

Number of students who took SI 697 or SI 698: 13


In [5]:
df["Course_Code"] = df["Course_Code"].astype(str).str.strip()
df["Course_Code"].value_counts()[["SI 699","SI 697","SI 698"]]
print("Number of students who took SI 699(Capstone), SI 697, or SI 698:")
print(df["Course_Code"].value_counts()[["SI 699","SI 697","SI 698"]])

Number of students who took SI 699(Capstone), SI 697, or SI 698:
Course_Code
SI 699    202
SI 697     13
SI 698     11
Name: count, dtype: int64


In [6]:
si699_students = set(df.loc[df["Course_Code"] == "SI 699", "Emplid"])
len(si699_students)

201

**SI 699 was taken 202 times, but by 201 people. So this means that one person took it twice.**

In [7]:
df[df["Course_Code"] == "SI 699"]["Emplid"].value_counts().loc[lambda x: x > 1]

Emplid
57381142    2
Name: count, dtype: int64

In [8]:
df[(df["Course_Code"] == "SI 699") & (df["Emplid"] == "57381142")]

,CompletionTermDesc,Emplid,Term_Descrshort,Crse_ID,Subject,Catalog_Nbr,Class_Section,Session_Descrshort,Crse_Component_Descrshort,Crse_Descr,Unt_Earned__Credits,Crse_Grade_Off__Grade_Earned,Course_Code,Outside_UMSI
1608,Winter 2022,57381142,WN 2021,47901,SI,699,13,Regular,Lecture,MSI Mastery Courses,3.0,B+,SI 699,No
1617,Winter 2022,57381142,WN 2022,47901,SI,699,3,Regular,Lecture,MSI Mastery Courses,3.0,A,SI 699,No


**Student 57381142 took SI 699 twice, in WN 2021 & WN 2022**

##### **Find MTOP Students**

In [9]:
mtop_students = set(df.loc[df["Course_Code"].isin(["SI 697","SI 698"]), "Emplid"].unique())
print(mtop_students,len(mtop_students))

{'68904555', '91993892', '43581370', '57767420', '14677197', '87410152', '52643901', '06526464', '46317563', '55887256', '50735760', '74249020', '66321952'} 13


##### **Find Students who have taken Neither (if applicable)**

In [10]:
missing_students = all_students - (si699_students | mtop_students)
missing_students, len(missing_students)

({'04388954', '35814158', '40239367'}, 3)

**Missing Students:**


04388954 - LAKES, 2013 Cohort


35814158 - BDA, Dual-Degree


40239367 - UX, Dual Degree

In [11]:
# Courseload for students who took neither SI 699 nor SI 697/698
df[df["Emplid"].isin(missing_students)].sort_values(["Emplid","Course_Code"])

,CompletionTermDesc,Emplid,Term_Descrshort,Crse_ID,Subject,Catalog_Nbr,Class_Section,Session_Descrshort,Crse_Component_Descrshort,Crse_Descr,Unt_Earned__Credits,Crse_Grade_Off__Grade_Earned,Course_Code,Outside_UMSI
3456,Fall 2022,04388954,FA 2022,13254,MUSICOL,503,1,Regular,Seminar,Mus Bibliogr,3.0,A,MUSICOL 503,Yes
3443,Fall 2022,04388954,WN 2014,42100,SI,500,1,Regular,Lecture,Info in Soc Systems,3.0,A,SI 500,No
3442,Fall 2022,04388954,FA 2013,42104,SI,501,1,Regular,Lecture,Contextual Inq,3.0,A,SI 501,No
3444,Fall 2022,04388954,FA 2014,42101,SI,502,6,Regular,Lecture,Net Computing,3.0,A,SI 502,No
3450,Fall 2022,04388954,FA 2018,38659,SI,519,1,Regular,Seminar,Intel Prop&Info Law,3.0,A,SI 519,No
3451,Fall 2022,04388954,WN 2019,41935,SI,530,1,Regular,Lecture,Prin in Management,3.0,A+,SI 530,No
3454,Fall 2022,04388954,FA 2021,9731,SI,580,1,Regular,Lecture,Records & Archives,3.0,A+,SI 580,No
3453,Fall 2022,04388954,WN 2021,50165,SI,585,1,Regular,Lecture,Scholarly Comm,3.0,A,SI 585,No
3445,Fall 2022,04388954,WN 2015,1451,SI,620,1,Regular,Lecture,Collection Dev&Mgt,3.0,A-,SI 620,No
3452,Fall 2022,04388954,WN 2020,39979,SI,623,1,Regular,Seminar,Res Meth Info Prof,3.0,P,SI 623,No


In [12]:
# Double Check
print("MTOP Students:", len(mtop_students))
print("Other Students:", len(missing_students))

MTOP Students: 13
Other Students: 3


In [13]:
# Combine MTOP and missing students to get the full list of students to exclude
exclude_students = mtop_students.union(missing_students)
print("Students to Exclude:", len(exclude_students))
exclude_students

Students to Exclude: 16


{'04388954',
 '06526464',
 '14677197',
 '35814158',
 '40239367',
 '43581370',
 '46317563',
 '50735760',
 '52643901',
 '55887256',
 '57767420',
 '66321952',
 '68904555',
 '74249020',
 '87410152',
 '91993892'}

##### **Narrow Dataset to Only Students Who Took SI 699**

In [14]:
# Filter Dataset to only include students who took SI 699
df_filtered = df[~df["Emplid"].isin(exclude_students)].copy()
df_filtered["Emplid"].nunique()

200

**Define Required Courses by Track**

In [15]:
# Course Pathway Course List by Required Classes Only
# Using only courses unique to each track as signals
# SI 582, SI 588, SI 622, SI 539 excluded from UX — shared with UCAD
# SI 501, SI 506 excluded from all — universal prereqs, not differentiating
track_required_courses = {
    "BDA":   ["SI 507", "SI 544", "SI 618", "SI 671"],
    "UX":    ["SI 520"],                                        # only UX-unique required course
    "UCAD":  ["SI 579", "SI 664", "SI 669"],                   # SI 579 is strongest UCAD signal
    "LAKES": ["SI 510", "SI 547", "SI 666", "SI 667",
              "SI 676", "SI 580", "SI 647"]
}

In [16]:
df_filtered = df_filtered.drop(columns=[c for c in df_filtered.columns if c.endswith("_flag")])

In [17]:
# Should show each flag column only once
print([c for c in df_filtered.columns if "flag" in c.lower()])

[]


**Basically checking if course is in each track's required list. If true, then the course is one of the track's required courses. Since some students could have taken a course for a different track as an elective, each course becomes a vote for the pathway. Whichever pathway gets the most votes will be the inferred pathway.**

In [18]:
# Course Pathway Course List by Required Classes Only
track_required_courses = {
    "BDA": ["SI 507", "SI 544", "SI 618", "SI 671"],
    "UX": ["SI 520"],
    "UCAD": ["SI 579", "SI 664", "SI 669"],
    "LAKES": ["SI 510", "SI 547", "SI 666", "SI 667", "SI 676", "SI 580", "SI 647"]
}

# Drop old flag columns if rerunning
df_filtered = df_filtered.drop(
    columns=[c for c in df_filtered.columns if c.lower().endswith("_flag")],
    errors="ignore"
)

# Create flag columns FIRST
for track, courses in track_required_courses.items():
    df_filtered[f"{track}_Flag"] = df_filtered["Course_Code"].isin(courses).astype(int)

# Then aggregate
track_counts = df_filtered.groupby("Emplid").agg(
    BDA_Count=("BDA_Flag", "sum"),
    UX_Count=("UX_Flag", "sum"),
    LAKES_Count=("LAKES_Flag", "sum"),
    UCAD_Count=("UCAD_Flag", "sum")
)

track_counts.head()

,BDA_Count,UX_Count,LAKES_Count,UCAD_Count
Emplid,,,,
00391172,2,1,0,0
01096243,1,0,0,2
01111805,1,0,0,0
01157522,1,0,0,0
01341800,0,1,0,0


**Summarize values for each flag, aka the number of "True"**

**Ex. Student 00391172 took 2 BDA required courses, 1 UX Required Courses,0 LAKES, and 0 UCAD. Will take the highest # and infer that the student is BDA.**

In [19]:
# Find students with zero required courses for any track
zero_track_students = track_counts[
    (track_counts["BDA_Count"] == 0) &
    (track_counts["UX_Count"] == 0) &
    (track_counts["LAKES_Count"] == 0) &
    (track_counts["UCAD_Count"] == 0)
]

print(zero_track_students)

print(len(zero_track_students), "students have zero required courses for any track.")

          BDA_Count  UX_Count  LAKES_Count  UCAD_Count
Emplid                                                
04743248          0         0            0           0
20144704          0         0            0           0
23497908          0         0            0           0
34179962          0         0            0           0
36208564          0         0            0           0
43571342          0         0            0           0
48554291          0         0            0           0
57777809          0         0            0           0
68685722          0         0            0           0
71425939          0         0            0           0
72056274          0         0            0           0
11 students have zero required courses for any track.


In [20]:
# Create indicator columns for whether each course is in the required courses for each track.
# Will later beummed per student to count how many required courses they took for each track.
for track, courses in track_required_courses.items():
    df_filtered[f"{track}_flag"] = df_filtered["Course_Code"].isin(courses)

# Use the counts of required courses ("votes") to infer the most likely track for each student.
# Calculate total votes, detect ties, and identify the track with the highest count.
vote_cols = ["BDA_Count", "UX_Count", "LAKES_Count", "UCAD_Count"]

# total votes + max vote per student
total_votes = track_counts[vote_cols].sum(axis=1)
max_vote = track_counts[vote_cols].max(axis=1)

# how many tracks tied for the max
number_at_max = track_counts[vote_cols].eq(max_vote, axis=0).sum(axis=1)

# best guess track
best_guess = (
    track_counts[vote_cols]
    .idxmax(axis=1)
    .str.replace("_Count", "", regex=False)
)

In [21]:
# Initialize all students as "Unknown", then assign tracks based on vote results.
track_counts["Inferred_Track"] = "Unknown"

# 0 votes → Zero-track (Students with no required-course signals are labeled "Zero-track".)
track_counts.loc[total_votes.eq(0), "Inferred_Track"] = "Zero-track"

# If multiple tracks share the highest vote count, label as "Tie".
tie_mask = total_votes.gt(0) & number_at_max.gt(1)
track_counts.loc[tie_mask, "Inferred_Track"] = "Tie"
# non-ties → assign best guess
assign_mask = total_votes.gt(0) & number_at_max.eq(1)
track_counts.loc[assign_mask, "Inferred_Track"] = best_guess[assign_mask]

track_counts["Inferred_Track"].value_counts()

Inferred_Track
BDA           62
LAKES         40
UX            34
Tie           33
UCAD          20
Zero-track    11
Name: count, dtype: int64

In [22]:
tie_ids = track_counts[track_counts["Inferred_Track"] == "Tie"].index
print(track_counts.loc[tie_ids, ["BDA_Count", "UX_Count", "LAKES_Count", "UCAD_Count"]].value_counts())

BDA_Count  UX_Count  LAKES_Count  UCAD_Count
1          1         0            0             6
0          1         1            0             6
1          1         0            1             6
2          0         0            2             4
0          1         0            1             4
3          0         0            3             2
           3         0            2             1
2          2         0            0             1
1          0         0            1             1
0          1         1            1             1
1          0         1            0             1
Name: count, dtype: int64


In [23]:
# Drop point columns if they already exist (makes cell safe to rerun)
point_cols = ["UX_Points", "UCAD_Points", "BDA_Points", "LAKES_Points"]
track_counts = track_counts.drop(columns=[c for c in point_cols if c in track_counts.columns])

# Comprehensive tiebreaker point system
UX_TIEBREAK    = ["SI 520", "SI 539", "SI 582", "SI 588", "SI 622"]
UCAD_TIEBREAK  = ["SI 579", "SI 664", "SI 669"]
BDA_TIEBREAK   = ["SI 608", "SI 630", "SI 649", "SI 650", "SI 670"]
LAKES_TIEBREAK = ["SI 623", "SI 530", "SI 632", "SI 633", "SI 638",
                  "SI 583", "SI 643", "SI 552", "SI 519"]

df_filtered["UX_Points"]    = df_filtered["Course_Code"].isin(UX_TIEBREAK).astype(int)
df_filtered["UCAD_Points"]  = df_filtered["Course_Code"].isin(UCAD_TIEBREAK).astype(int)
df_filtered["BDA_Points"]   = df_filtered["Course_Code"].isin(BDA_TIEBREAK).astype(int)
df_filtered["LAKES_Points"] = df_filtered["Course_Code"].isin(LAKES_TIEBREAK).astype(int)

points = df_filtered.groupby("Emplid")[
    ["UX_Points", "UCAD_Points", "BDA_Points", "LAKES_Points"]
].sum()

track_counts = track_counts.join(points, how="left")
track_counts[["UX_Points", "UCAD_Points", "BDA_Points", "LAKES_Points"]] = (
    track_counts[["UX_Points", "UCAD_Points", "BDA_Points", "LAKES_Points"]].fillna(0)
)

In [24]:
# Reassign tie students using all four track signals
tie_mask = track_counts["Inferred_Track"].isin(["Tie", "Manual"])

point_cols = ["BDA_Points", "UX_Points", "LAKES_Points", "UCAD_Points"]
col_to_track = {
    "BDA_Points":   "BDA",
    "UX_Points":    "UX",
    "LAKES_Points": "LAKES",
    "UCAD_Points":  "UCAD"
}

for idx in track_counts[tie_mask].index:
    row = track_counts.loc[idx, point_cols]
    max_pts = row.max()
    if max_pts == 0 or (row == max_pts).sum() > 1:
        track_counts.loc[idx, "Inferred_Track"] = "Tie"
    else:
        track_counts.loc[idx, "Inferred_Track"] = col_to_track[row.idxmax()]

track_counts["Inferred_Track"].value_counts()

Inferred_Track
BDA           67
UX            60
LAKES         40
UCAD          21
Zero-track    11
Tie            1
Name: count, dtype: int64

## **IN-DEPTH ANALYSIS ON TIES**

In [25]:
tie_ids = track_counts[track_counts["Inferred_Track"] == "Tie"].index
print(f"{len(tie_ids)} unresolvable tie students:\n")
for emplid in tie_ids:
    courses = sorted(df_filtered[df_filtered["Emplid"] == emplid]["Course_Code"].unique())
    print(f"{emplid}: {courses}")
    print(track_counts.loc[emplid, ["BDA_Count", "UX_Count", "LAKES_Count", "UCAD_Count",
                                     "BDA_Points", "UX_Points", "LAKES_Points", "UCAD_Points"]])
    print()

1 unresolvable tie students:

02585323: ['LHS 712', 'PUBPOL 650', 'SI 501', 'SI 504', 'SI 539', 'SI 564', 'SI 582', 'SI 608', 'SI 612', 'SI 617', 'SI 618', 'SI 664', 'SI 669', 'SI 670', 'SI 671', 'SI 681', 'SI 699']
BDA_Count       2
UX_Count        0
LAKES_Count     0
UCAD_Count      2
BDA_Points      2
UX_Points       2
LAKES_Points    0
UCAD_Points     2
Name: 02585323, dtype: int64



In [26]:
df_filtered[df_filtered["Emplid"].isin(
    track_counts[track_counts["Inferred_Track"] == "Tie"].index
)]

,CompletionTermDesc,Emplid,Term_Descrshort,Crse_ID,Subject,Catalog_Nbr,Class_Section,Session_Descrshort,Crse_Component_Descrshort,Crse_Descr,...,UCAD_Flag,LAKES_Flag,BDA_flag,UX_flag,UCAD_flag,LAKES_flag,UX_Points,UCAD_Points,BDA_Points,LAKES_Points
186,Winter 2022,02585323,FA 2020,41934,SI,608,1,Regular,Lecture,Networks,...,0,0,False,False,False,False,0,0,1,0
187,Winter 2022,02585323,FA 2020,42091,SI,618,101,Regular,Lecture,Data Man & Analysis,...,0,0,True,False,False,False,0,0,0,0
188,Winter 2022,02585323,FA 2020,42104,SI,501,1,Regular,Lecture,Contextual Inq,...,0,0,False,False,False,False,0,0,0,0
189,Winter 2022,02585323,FA 2020,44772,SI,582,4,Regular,Lecture,Intro Interac Design,...,0,0,False,False,False,False,1,0,0,0
190,Winter 2022,02585323,WN 2021,39178,SI,539,1,Regular,Lecture,Des of Comp Websites,...,0,0,False,False,False,False,1,0,0,0
191,Winter 2022,02585323,WN 2021,40667,SI,681,89,Regular,Seminar,Intern/Field Seminar,...,0,0,False,False,False,False,0,0,0,0
192,Winter 2022,02585323,WN 2021,41736,PUBPOL,650,1,Regular,Lecture,Intr to Science&Tech,...,0,0,False,False,False,False,0,0,0,0
193,Winter 2022,02585323,WN 2021,43366,SI,617,1,Regular,Lecture,Choice Architecture,...,0,0,False,False,False,False,0,0,0,0
194,Winter 2022,02585323,WN 2021,43863,SI,612,1,Regular,Lecture,Pervasv Interact Des,...,0,0,False,False,False,False,0,0,0,0
195,Winter 2022,02585323,WN 2021,47637,LHS,712,1,Regular,Lecture,NLP on Health Data,...,0,0,False,False,False,False,0,0,0,0


Based on the classes Student 02585323 has taken, I have determined that the student is part of the BDA track. That student has been manually assigned as such.

In [27]:
track_counts.loc[track_counts["Inferred_Track"] == "Tie", "Inferred_Track"] = "BDA"

track_counts["Inferred_Track"].value_counts()

Inferred_Track
BDA           68
UX            60
LAKES         40
UCAD          21
Zero-track    11
Name: count, dtype: int64

## **IN-DEPTH ANALYSIS ON ZERO-TRACK**

In [28]:
# Identify students with zero required-course signals and list all courses they took
zero_ids = track_counts[track_counts["Inferred_Track"] == "Zero-track"].index
for emplid in zero_ids:
    courses = sorted(df_filtered[df_filtered["Emplid"] == emplid]["Course_Code"].unique())
    print(f"{emplid}: {courses}")
    print()

04743248: ['BA 670', 'ENTR 599', 'MO 637', 'SI 501', 'SI 506', 'SI 519', 'SI 529', 'SI 539', 'SI 552', 'SI 582', 'SI 588', 'SI 612', 'SI 622', 'SI 658', 'SI 681', 'SI 699']

20144704: ['BA 670', 'ENTR 500', 'ENTR 560', 'SI 501', 'SI 559', 'SI 563', 'SI 649', 'SI 690', 'SI 699']

23497908: ['ENTR 550', 'ENTR 560', 'SI 501', 'SI 537', 'SI 538', 'SI 627', 'SI 682', 'SI 699']

34179962: ['ARCH 509', 'ENTR 560', 'GREEKMOD 502', 'SI 501', 'SI 563', 'SI 612', 'SI 682', 'SI 690', 'SI 699']

36208564: ['DESCI 501', 'ENTR 500', 'SI 505', 'SI 539', 'SI 554', 'SI 564', 'SI 588', 'SI 658', 'SI 681', 'SI 682', 'SI 699']

43571342: ['BA 670', 'IOE 548', 'SI 501', 'SI 506', 'SI 538', 'SI 539', 'SI 559', 'SI 582', 'SI 588', 'SI 612', 'SI 616', 'SI 622', 'SI 681', 'SI 699', 'SPANISH 101', 'TO 548']

48554291: ['AMCULT 658', 'PUBPOL 650', 'SI 501', 'SI 519', 'SI 582', 'SI 588', 'SI 690', 'SI 699']

57777809: ['EDUC 602', 'EDUC 616', 'EDUC 617', 'SI 501', 'SI 506', 'SI 519', 'SI 539', 'SI 549', 'SI 559', 

In [29]:
# Look at their most common courses
zero_ids = track_counts[track_counts["Inferred_Track"] == "Zero-track"].index
zero_courses = df_filtered[df_filtered["Emplid"].isin(zero_ids)]

print("Most common courses among zero-track students:")
print(zero_courses["Course_Code"].value_counts().head(20))

Most common courses among zero-track students:
Course_Code
SI 699      11
SI 501      10
SI 690       8
SI 588       6
SI 658       6
SI 681       6
SI 582       5
SI 539       5
SI 612       5
SI 506       4
SI 622       4
SI 519       4
SI 559       4
SI 682       4
SI 552       3
BA 670       3
ENTR 560     3
SI 538       3
SI 505       3
SI 563       2
Name: count, dtype: int64


In [30]:
zero_track_signals = {
    "UX":    ["SI 520", "SI 539", "SI 582", "SI 588", "SI 622"],
    "UCAD":  ["SI 579", "SI 664", "SI 669"],
    "BDA":   ["SI 608", "SI 630", "SI 649", "SI 650", "SI 670"],
    "LAKES": ["SI 510", "SI 547", "SI 666", "SI 667", "SI 676",
              "SI 580", "SI 647", "SI 530", "SI 552", "SI 538"]
}

for track, courses in zero_track_signals.items():
    zero_courses_df = df_filtered[df_filtered["Emplid"].isin(zero_ids)].copy()
    
soft_votes = df_filtered[df_filtered["Emplid"].isin(zero_ids)].copy()

for track, courses in zero_track_signals.items():
    soft_votes[f"{track}_soft"] = soft_votes["Course_Code"].isin(courses).astype(int)

soft_votes = soft_votes.groupby("Emplid")[
    ["UX_soft", "UCAD_soft", "BDA_soft", "LAKES_soft"]
].sum()

print(soft_votes)

          UX_soft  UCAD_soft  BDA_soft  LAKES_soft
Emplid                                            
04743248        4          0         0           1
20144704        0          0         1           0
23497908        0          0         0           1
34179962        0          0         0           0
36208564        2          0         0           0
43571342        4          0         0           1
48554291        2          0         0           0
57777809        4          0         0           0
68685722        0          0         0           2
71425939        4          0         0           1
72056274        0          0         0           2


In [31]:
# Take closer look at students with 0 soft signals by looking up a specific student's track and vote counts
for emplid in ["23497908", "34179962"]:
    print(f"\n{'='*50}")
    print(f"Student: {emplid}")
    print(f"\nCourses taken:")
    print(sorted(df_filtered[df_filtered["Emplid"] == emplid]["Course_Code"].unique()))
    print(f"\nVote counts:")
    print(track_counts.loc[emplid, ["BDA_Count", "UX_Count", "LAKES_Count", "UCAD_Count",
                                     "BDA_Points", "UX_Points", "LAKES_Points", "UCAD_Points",
                                     "Inferred_Track"]])


Student: 23497908

Courses taken:
['ENTR 550', 'ENTR 560', 'SI 501', 'SI 537', 'SI 538', 'SI 627', 'SI 682', 'SI 699']

Vote counts:
BDA_Count                  0
UX_Count                   0
LAKES_Count                0
UCAD_Count                 0
BDA_Points                 0
UX_Points                  0
LAKES_Points               0
UCAD_Points                0
Inferred_Track    Zero-track
Name: 23497908, dtype: object

Student: 34179962

Courses taken:
['ARCH 509', 'ENTR 560', 'GREEKMOD 502', 'SI 501', 'SI 563', 'SI 612', 'SI 682', 'SI 690', 'SI 699']

Vote counts:
BDA_Count                  0
UX_Count                   0
LAKES_Count                0
UCAD_Count                 0
BDA_Points                 0
UX_Points                  0
LAKES_Points               0
UCAD_Points                0
Inferred_Track    Zero-track
Name: 34179962, dtype: object


In [32]:
# Assign zero-track students based on soft votes
soft_col_to_track = {"UX_soft": "UX", "UCAD_soft": "UCAD",
                     "BDA_soft": "BDA", "LAKES_soft": "LAKES"}

soft_assignments = {}
for emplid, row in soft_votes.iterrows():
    max_votes = row.max()
    if max_votes == 0 or (row == max_votes).sum() > 1:
        soft_assignments[emplid] = "Manual"
    else:
        soft_assignments[emplid] = soft_col_to_track[row.idxmax()]

# Manual overrides
soft_assignments["23497908"] = "UX"
soft_assignments["34179962"] = "UX"

print(soft_assignments)

{'04743248': 'UX', '20144704': 'BDA', '23497908': 'UX', '34179962': 'UX', '36208564': 'UX', '43571342': 'UX', '48554291': 'UX', '57777809': 'UX', '68685722': 'LAKES', '71425939': 'UX', '72056274': 'LAKES'}


In [33]:
# Apply soft assignments to track_counts
for emplid, track in soft_assignments.items():
    track_counts.loc[emplid, "Inferred_Track"] = track

# Final check — should show 0 Zero-track
print(track_counts["Inferred_Track"].value_counts())
assert "Zero-track" not in track_counts["Inferred_Track"].values, "Still have unassigned students!"
assert len(track_counts) == 200, "Should have exactly 200 students!"

Inferred_Track
BDA      69
UX       68
LAKES    42
UCAD     21
Name: count, dtype: int64


## **ELECTIVE ANALYSIS**

In [34]:
# Re-merge finalized track labels into course dataset
track_labels = track_counts.reset_index()[["Emplid", "Inferred_Track"]]
df_with_tracks = df_filtered.merge(track_labels, on="Emplid", how="left")

print(df_with_tracks["Inferred_Track"].value_counts())
print(f"Unique students: {df_with_tracks['Emplid'].nunique()}")  # should be 200

Inferred_Track
BDA      1116
UX       1106
LAKES     662
UCAD      325
Name: count, dtype: int64
Unique students: 200


In [35]:
# Required courses per track + universal exclusions
REQUIRED = {
    "BDA":   {"SI 507", "SI 544", "SI 618", "SI 671"},
    "UX":    {"SI 520", "SI 539", "SI 582", "SI 588", "SI 622"},  # added shared prereqs
    "UCAD":  {"SI 579", "SI 664", "SI 669", "SI 539", 
              "SI 582", "SI 588", "SI 622"},                       # added shared prereqs
    "LAKES": {"SI 510", "SI 547", "SI 666", "SI 667",
              "SI 676", "SI 580", "SI 647"}
}

# Universal exclusions: prereqs, internship, capstone
UNIVERSAL_EXCLUDE = {"SI 501", "SI 505", "SI 506",
                     "SI 681", "SI 690", "SI 699"}

In [36]:
results = {}
for track in ["BDA", "UX", "UCAD", "LAKES"]:
    exclude = UNIVERSAL_EXCLUDE | REQUIRED[track]
    n_students = (track_counts["Inferred_Track"] == track).sum()

    track_df = df_with_tracks[
        (df_with_tracks["Inferred_Track"] == track) &
        (~df_with_tracks["Course_Code"].isin(exclude))
    ]

    elective_counts = (
        track_df.groupby(["Course_Code", "Crse_Descr"])["Emplid"]
        .nunique()
        .sort_values(ascending=False)
        .rename("Student_Count")
        .reset_index()
    )

    elective_counts["Pct_of_Track"] = (
        elective_counts["Student_Count"] / n_students * 100
    ).round(1)

    results[track] = elective_counts

    print(f"\n── {track} (n={n_students} students) ──────────────────")
    print(elective_counts.head(10).to_string(index=False))


── BDA (n=69 students) ──────────────────
Course_Code           Crse_Descr  Student_Count  Pct_of_Track
     SI 670 Appld Machine Learng             41          59.4
     SI 564      SQL & Databases             40          58.0
     SI 649   Info Visualization             39          56.5
     SI 664  Database App Design             39          56.5
     SI 650  Informatn Retrieval             34          49.3
     SI 630 NLP Algorithm People             28          40.6
     SI 539 Des of Comp Websites             24          34.8
     SI 582 Intro Interac Design             24          34.8
     SI 504  Servers, Shell,&Git             23          33.3
     SI 588   Fundamntls Hum Beh             17          24.6

── UX (n=68 students) ──────────────────
Course_Code           Crse_Descr  Student_Count  Pct_of_Track
     SI 612 Pervasv Interact Des             35          51.5
     SI 559          AR/VR Intro             33          48.5
     SI 616   Adv Graphic Design             27

In [37]:
print(df_with_tracks["Outside_UMSI"].value_counts())

Outside_UMSI
No     2813
Yes     396
Name: count, dtype: int64


In [38]:
outside_results = {}

for track in ["BDA", "UX", "UCAD", "LAKES"]:
    exclude = UNIVERSAL_EXCLUDE | REQUIRED[track]
    n_students = (track_counts["Inferred_Track"] == track).sum()

    outside_df = df_with_tracks[
        (df_with_tracks["Inferred_Track"] == track) &
        (~df_with_tracks["Course_Code"].isin(exclude)) &
        (df_with_tracks["Outside_UMSI"] == "Yes")
    ]

    outside_counts = (
        outside_df.groupby(["Course_Code", "Crse_Descr"])["Emplid"]
        .nunique()
        .sort_values(ascending=False)
        .rename("Student_Count")
        .reset_index()
    )
    outside_counts["Pct_of_Track"] = (
        outside_counts["Student_Count"] / n_students * 100
    ).round(1)

    outside_results[track] = outside_counts
    print(f"\n── {track} (n={n_students} students) ──────────────────")
    print(outside_counts.head(10).to_string(index=False))


── BDA (n=69 students) ──────────────────
Course_Code           Crse_Descr  Student_Count  Pct_of_Track
    LHS 712   NLP on Health Data              5           7.2
   ENTR 560 Proj Mgmt Consulting              5           7.2
   ENTR 500  Intro to Innovation              5           7.2
     MO 512  Bargain&Infl Skills              4           5.8
   ENGR 599     Spec Topics Engr              3           4.3
   ENTR 550 Interpersonal Skills              3           4.3
   EECS 598       Special Topics              3           4.3
   EECS 592       AI Foundations              3           4.3
   EECS 586           Algorithms              3           4.3
 PUBPOL 650 Intr to Science&Tech              3           4.3

── UX (n=68 students) ──────────────────
Course_Code           Crse_Descr  Student_Count  Pct_of_Track
   ENTR 500  Intro to Innovation             10          14.7
   EDUC 617      XR for Learning              9          13.2
     MO 512  Bargain&Infl Skills              7

In [39]:
# Export full elective list for all tracks — for student-facing course directory
import json

full_course_map = {}

for track, df_result in results.items():
    # Only include SI courses (not outside UMSI)
    si_only = df_result[df_result['Course_Code'].str.startswith('SI ')]
    for _, row in si_only.iterrows():
        code = row['Course_Code']
        name = row['Crse_Descr']
        if code not in full_course_map:
            full_course_map[code] = {'name': name, 'tracks': []}
        if track not in full_course_map[code]['tracks']:
            full_course_map[code]['tracks'].append(track)

# Print as JavaScript array ready to paste into your HTML file
print('const allCourses = [')
for code in sorted(full_course_map.keys(), key=lambda x: int(x.split()[1])):
    info = full_course_map[code]
    tracks = sorted([t.lower() for t in info['tracks']])
    tracks_str = ', '.join(f"'{t}'" for t in tracks)
    name = info['name'].replace("'", "\\'")
    print(f"  {{ code: '{code}', name: '{name}', tracks: [{tracks_str}] }},")
print('];')

const allCourses = [
  { code: 'SI 504', name: 'Servers, Shell,&Git', tracks: ['bda', 'lakes', 'ucad', 'ux'] },
  { code: 'SI 507', name: 'Inter Programming', tracks: ['lakes', 'ucad', 'ux'] },
  { code: 'SI 508', name: 'Intensive Info Prog', tracks: ['bda'] },
  { code: 'SI 510', name: 'Information Ethics', tracks: ['bda', 'ux'] },
  { code: 'SI 511', name: 'Special Topics', tracks: ['bda', 'ux'] },
  { code: 'SI 512', name: 'Change The Internet', tracks: ['ux'] },
  { code: 'SI 519', name: 'Intel Prop&Info Law', tracks: ['bda', 'lakes', 'ux'] },
  { code: 'SI 520', name: 'Graphic Design', tracks: ['bda', 'lakes', 'ucad'] },
  { code: 'SI 529', name: 'Online Communities', tracks: ['bda', 'lakes', 'ucad', 'ux'] },
  { code: 'SI 530', name: 'Prin in Management', tracks: ['bda', 'lakes', 'ucad', 'ux'] },
  { code: 'SI 534', name: 'Theories Soc Influen', tracks: ['bda', 'lakes', 'ux'] },
  { code: 'SI 536', name: 'Intro to Urbn Inform', tracks: ['bda'] },
  { code: 'SI 537', name: 'Crisis